### ISSUE TITLE
Javascript client array not deserializing on Server

### ISSUE THREAD

**Issue by LukeHartcher** [DataBooster/DbWebApi#1]
I am trying to pass a JS data array through to the server for bulk insertion but I receive an deserialize error. See the details/examples below.

#### Javascript Array Construction
I am constructing my JS array as follows:
```
var dataParams = [];
// foreach item...
dataParams.push({
                ID1: 1,
                ID2: 2,
                Value: 'xyz'
            });

$.getDb(url, dataParams, function(response){
   /// do something
});
```

#### ERROR
I receive this error in .Net:
```
Cannot deserialize the current JSON array (e.g. [1,2,3]) into type 'System.Collections.Generic.Dictionary`2[System.String,System.Object]' because the type requires a JSON object (e.g. {"name":"value"}) to deserialize correctly.

To fix this error either change the JSON to a JSON object (e.g. {"name":"value"}) or change the deserialized type to an array or a type that implements a collection interface (e.g. ICollection, IList) like List<T> that can be deserialized from a JSON array. JsonArrayAttribute can also be added to the type to force it to deserialize from a JSON array.
```

**Comment by AbelCheng** [comment]
[BulkExecute reads bulk sets of parameters from the request message body only, it means only HTTP **POST** and PUT can be used to send BulkExecute request.](https://github.com/DataBooster/DbWebApi/wiki#array-of-parameter-sets)
So, the method "$.getDb(...)" in your example code should be changed to "**$.postDb**(...)", then it will work.

BTW, According to your JS example, your database side stored procedure should be a regular single row insert (or update) statement. This could be effective enough if each time you don't have too many records to write back to database.  
However, if you do have lots of data records (e.g. hundreds or thousands of rows) to write back to database at each time, please consider using [Table-Valued Parameters (SQL Server 2008+)](https://github.com/DataBooster/DbWebApi/wiki#table-valued-parameters) or [PL/SQL Associative Array Parameters (Oracle)](https://github.com/DataBooster/DbWebApi/wiki#associative-array-parameters) features, that would be much faster.

Hope these help.

**Comment by LukeHartcher** [comment]
Thanks for the prompt response.

The `$.postDb(...)` call does not work with a different error on the server:

```
Procedure or function 'myProc' expects parameter '@myParam', which was not supplied.
```

The `allParameters = InputParameters.SupplementFromQueryString(allParameters, Request);` on the WebApi end is not returning any data so `allParameters.Parameters.Count = 0`.

I tried a manual `$.ajax()` post with `data=dataParams` and `dataType=json` but I get the same missing parameter issue as with the `$.postDb()`.

Any thoughts? Am I not passing the array through correctly? Am I missing some code in my DbWebApiController (which I copied from your sample).

**Comment by AbelCheng** [comment]
To help addressing the issue, need some more information from you:

1. Could you capture (e.g. Fiddler, or just IE F12 Developer Tools - Network) the HTTP traffic? I would like to see what exact content was sent in the HTTP request **body**.

2. If you don't mind, may I know the **signature of your stored procedure**? Make sure every input parameter name can get match from the JSON request.

3. Could you do another quick test from your browser, type "http://your-base-url/dbname.dbo.myProc?ID1=1&ID2=2&Value=xyz" in the address bar, just to check the server side is working or not.

4. While you debugging the server side code, please check the allParameters.**Bulk**Parameters property instead of allParameters.Parameters. Inside the **BulkParameters**, you should see a list/array of Parameters, check each of them. Because in your case, the client side is sending an array of parameter-set. To check the server side has been switched to BulkExecute mode or not, allParameters.**ForBulkExecuting** property indicates that (you should see it = **true** in your case).

Hope to hear your progress soon.

**Comment by LukeHartcher** [comment]
Still no luck. As this is not working I do a "singular" call to the stored proc in my forEach and this works fine.

Here is some code/data as per your request. Hopefully you can find out where I am going wrong.

I appreciate your assistance, many thanks!!

### Javascript Code
```
// dataParrams Object. Need to wrap array in Json Object otherwise .Net wont parse it at all.
var dataParams = { array: [] };
$rootScope.globalModalEditData.dbMeasurements.forEach(function(measurement){
	dataParams.array.push({
		ContactNo: measurement.ContactNo,
		MeasurementTypeID: measurement.MeasurementTypeID,
		Measurement: measurement.Measurement
	});
});
$.ajax({
	type: 'post',
	url: appConfig.dbWebApiUrl + appConfig.dbWebApiDbase + '.sproc_InsertUpdateContactMeasurement/json',
	dataType: 'json',
	data: JSON.stringify(dataParams)
}).success(function (successResponse) {
	$log.debug('appData.ajaxDb: ' + storedProc + ' SUCCESS', successResponse);
	if (onComplete) {
		$timeout(function () {
			onComplete(response);
		});
	}
}).fail(function (errorResponse) {
	$log.debug('appData.ajaxDb: ' + storedProc + ' ERROR', errorResponse);
});
```

### Request Object via Fiddler
```
POST http://localhost:54404/dbWebApi/SisleyClothing.dbo.sproc_InsertUpdateContactMeasurement/json HTTP/1.1
Host: localhost:54404
Connection: keep-alive
Content-Length: 429
Pragma: no-cache
Cache-Control: no-cache
Accept: application/json, text/javascript, */*; q=0.01
Origin: http://192.168.10.154:3040
User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.112 Safari/537.36
Content-Type: application/x-www-form-urlencoded; charset=UTF-8
Referer: http://192.168.10.154:3040/
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.8,zh-CN;q=0.6,zh;q=0.4

{"array":[{"ContactNo":2,"MeasurementTypeID":1,"Measurement":"123abc"},{"ContactNo":2,"MeasurementTypeID":2,"Measurement":"789"},{"ContactNo":2,"MeasurementTypeID":3,"Measurement":"456def"},{"ContactNo":2,"MeasurementTypeID":4,"Measurement":"something"},{"ContactNo":2,"MeasurementTypeID":5,"Measurement":"631"},{"ContactNo":2,"MeasurementTypeID":6,"Measurement":"654"},{"ContactNo":2,"MeasurementTypeID":7,"Measurement":"479"}]}
```

### Stored Procedure Header
```
ALTER PROCEDURE [dbo].[sproc_InsertUpdateContactMeasurement]
(
	@ContactNo int,
	@MeasurementTypeID int,
	@Measurement varchar(50)
)
AS
```

### .Net WebApi Controller Code
```
[AcceptVerbs("GET", "POST", "PUT", "DELETE", "OPTIONS")]
public HttpResponseMessage DynExecute(string sp, InputParameters allParameters)
{
	// Supplement input parameters from URI query string.
	allParameters = InputParameters.SupplementFromQueryString(allParameters, Request);

	// The main entry point to call the DbWebApi.
	return this.DynExecuteDbApi(sp, allParameters);
}
```

**Comment by LukeHartcher** [comment]
I was able to get it working - **but there is appears to be .Net JSON Issue!** (FYI - tried .Net 4.5 & 4.6.1)

Essentially the JSON is getting corrupt by .Net trying to create the `IDictionary<string, object>` object it passes to the `DynExecute` controller action - meaning the JSON is malformed.

If we FIX the JSON and then rebuild the allParameters via a new `allParameters = new InputParameters(JArray.Parse(validJson))` call it all works!

I added the function below to my controller, but perhaps you need to cater for this? Or am I just doing something completely wrong in the first place that is causing .Net to build the IDictionary incorrectly?

```
private bool fixJsonArray(InputParameters inputParameters, out InputParameters newParameters)
{
	var didFix = false;
	newParameters = null;
	try
	{
		// If we have a and single param check to see if it contains an IDictionary 
		if (inputParameters != null && inputParameters.Parameters != null && inputParameters.Parameters.Count == 1)
		{
			Object firstObj = inputParameters.Parameters.First().Value;
			if(firstObj is IDictionary<string, object>)
			{
				// The Key will contain the 'broken' JSON string we are after
				string corruptJson = (firstObj as IDictionary < string, object>).First().Key;
				// Fix Json string - put back in the open array '[' and remove the ending '}' from the intial JSON object.
				string validJson = "[" + corruptJson.Substring(0, corruptJson.Length - 1);
				// Use a JArray to recreate the InputParameters
				newParameters = new InputParameters(JArray.Parse(validJson));
				didFix = true;
			}
		}
	}
	catch (Exception)
	{
		// do nothing
	}
	return didFix;
}
```

and then in my `DynExecute` controller action

```
InputParameters newParameters;
if( fixJsonArray(allParameters, out newParameters) )
{
	allParameters = newParameters;
}
```

**Comment by AbelCheng** [comment]
Did/could you try your first approach (change $.getDb to $.postDb), and restore server side .net code as well? To capture the request message via Fiddler here.
```
var dataParams = new Array();
// foreach item...
dataParams.push({
                  ContactNo: 2,
                  MeasurementTypeID: 1,
                  Measurement: '123abc'
                });

$.postDb(url, dataParams, function(response){
   /// do something
});
```
My expected HTTP request message should be:
```
Accept: application/json, */*
Content-Type: application/json

[{"ContactNo":2,"MeasurementTypeID":1,"Measurement":"123abc"},{"ContactNo":2,"MeasurementTypeID":2,"Measurement":"789"},{"ContactNo":2,"MeasurementTypeID":3,"Measurement":"456def"},{"ContactNo":2,"MeasurementTypeID":4,"Measurement":"something"},{"ContactNo":2,"MeasurementTypeID":5,"Measurement":"631"},{"ContactNo":2,"MeasurementTypeID":6,"Measurement":"654"},{"ContactNo":2,"MeasurementTypeID":7,"Measurement":"479"}]
```
Because the server side operation modes (singular execution or bulk execution) is determined by the client's request JSON:
- If the whole body looks like `{...}` (the outermost level is `{` `}`), then InputParameters will be switched to singular execution mode;
```
   allParameters.ForBulkExecuting = false
   allParameters.Parameters = A dictionary of sp parameters
   allParameters.BulkParameters = null
```
- If the whole body looks like `[{...},{...},...]` (the outermost level is `[` `]`), then InputParameters will be switched to bulk execution mode;
```
   allParameters.ForBulkExecuting = true
   allParameters.Parameters = null
   allParameters.BulkParameters = An array of dictionaries of sp parameters
```
For trouble shooting this issue, I prefer to make sure the HTTP request message sent from the client side can match how design works in the server side first.

**Comment by LukeHartcher** [comment]
I think I have a CORS issue with my server code.

As the post content type is "application/json" (and not "application/x-www-form-urlencoded") it is kicking off the pre-flight OPTIONS post. This is why I thought I had no data on the server, but the actual POST post is not getting to the server as the OPTIONS post is failing.

I will look into my CORS implementation to see if I can resolve it.

Will let you know how I get on, thanks.

**Comment by LukeHartcher** [comment]
**FIXED**

My issue was that I was not handling the **CORS OPTIONS** Request which gets generated when the `contentType` is changed to `application/json` (from `application/x-www-form-urlencoded` which does not require CORS). This is why I was seeing the empty data, as this was actually the OPTIONS post I was seeing as the controller `DynExecute` Action was set to accept OPTIONS requests.

To resolve I did two things:
1. Handled the **OPTIONS** request in my `Global.asax.cs` (simply returning an empty response)
2. Removed the `"OPTIONS"` Verb from the controller `DynExecute` Action so that the OPTIONS was not reaching the Controller.

Now I can set my array back to the correct format and I have **removed** my `fixJsonArray` routine as I no longer require it.

Fiddler is my friend! :)
